xmtp_mls/src/groups/mod.rs

```diff
@@ -522,8 +541,44 @@ where
     // Update this installation's leaf key in the group by creating a key update commit
     pub async fn key_update(&self) -> Result<(), GroupError> {
         let conn = &mut self.client.store.conn()?;
+        
+        self.pre_intent_hook().await.unwrap();
+
+        let intent = NewGroupIntent::new(IntentKind::KeyUpdate, self.group_id.clone(), vec![]);
+        intent.store(conn)?;
+
+        self.sync_with_conn(conn).await
+    }
+
+    // Update the installation's leaf key in a synchronous way.
+    pub fn queue_key_update(&self) -> Result<(), GroupError> {
+        let conn = &mut self.client.store.conn()?;
+        
+        self.pre_intent_hook();
```

Comment:
There's an infinite loop here because `queue_key_update` and `pre_intent_hook` both call each other.

What we can do is *not* call `pre_intent_hook` from `queue_key_update`. That way, `queue_key_update` can only be reached from `pre_intent_hook` but not the other way around. Then you can remove the `pub` from the signature of `queue_key_update`.

Also, I'm not sure if we need both `queue_key_update` and `key_update`, it seems like only one of the two is sufficient.

Reply:
It might make sense to write a unit test that reproduces this, before fixing it, to make sure that the fix works

Reply:
> Also, I'm not sure if we need both queue_key_update and key_update, it seems like only one of the two is sufficient.

I think it might make sense to keep `key_update` and remove `queue_key_update`, seeing as you no longer have the 'calling async method from inside sync method' issue. You don't need to update the key inside of `create_from_welcome` anymore, because the key update will happen on the next message send.

You can call `key_update` from inside of `pre_intent_hook`, but you'll need to remove `pre_intent_hook` from inside of `key_update` too